### ISSUE TITLE
[FSDP] caffe2 error in forward method when using fsdp

### ISSUE THREAD

**Issue by pacman100** [pytorch/pytorch#82461]
### 🐛 Describe the bug

When using FSDP, during inference/evaluation using transformers (gpt2, blenderbot, t5 ...) for generation, i.e., `model.generate()`, caffe2 error is thrown.

### Steps to reproduce the error:
1. Code is here: [run_seq2seq_no_trainer.py](https://github.com/pacman100/fsdp-tests/blob/main/src/modeling/run_seq2seq_no_trainer.py)
2. Using 🤗 Accelerate's FSDP integration with [config.yaml](https://github.com/pacman100/fsdp-tests/blob/main/src/modeling/config.yaml) being below.  Using 2 Nvidia Titan RTX GPUs. 
```yaml
compute_environment: LOCAL_MACHINE
deepspeed_config: {}
distributed_type: FSDP
downcast_bf16: 'no'
fsdp_config:
  fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
  fsdp_backward_prefetch_policy: BACKWARD_PRE
  fsdp_offload_params: false
  fsdp_sharding_strategy: 1
  fsdp_state_dict_type: FULL_STATE_DICT
  fsdp_transformer_layer_cls_to_wrap: BlenderbotDecoderLayer
machine_rank: 0
main_process_ip: null
main_process_port: null
main_training_function: main
mixed_precision: 'fp16'
num_machines: 1
num_processes: 2
use_cpu: false
```
3. Running below [launch](https://github.com/pacman100/fsdp-tests/blob/main/src/modeling/seq2seq_script.sh) command:
```bash
accelerate launch --config_file config.yaml \
    run_seq2seq_no_trainer.py \
    --dataset_name "smangrul/MuDoConv" \
    --max_source_length 128 \
    --source_prefix "chatbot: " \
    --max_target_length 64 \
    --val_max_target_length 64 \
    --val_min_target_length 20 \
    --n_val_batch_generations 5 \
    --n_train 10000 \
    --n_val 1000 \
    --pad_to_max_length \
    --num_beams 10 \
    --model_name_or_path "facebook/blenderbot-400M-distill" \
    --per_device_train_batch_size 100 \
    --per_device_eval_batch_size 50 \
    --learning_rate 1e-6 \
    --weight_decay 0.0 \
    --num_train_epochs 1 \
    --gradient_accumulation_steps 1 \
    --num_warmup_steps 100 \
    --output_dir "/tmp/fsdp_test" \
    --seed 25 \
    --logging_steps 100
```
4. Output with the error and trace stack:
```bash
Traceback (most recent call last):                                                                                             
  File "run_seq2seq_no_trainer.py", line 911, in <module>                                                                      
    bleu_score = evaluate(args, model, metric, tokenizer, eval_dataloader, accelerator, config.max_length)                     
  File "run_seq2seq_no_trainer.py", line 392, in evaluate                                                                      
    generated_tokens = unwrapped_model.generate(                                                                               
  File "/home/sourab/dev/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context                
    main()                                                                                                                     
  File "run_seq2seq_no_trainer.py", line 840, in main                                                                          
    return func(*args, **kwargs)                                                                                               
  File "/home/sourab/transformers/src/transformers/generation_utils.py", line 1181, in generate                                
    bleu_score = evaluate(args, model, metric, tokenizer, eval_dataloader, accelerator, config.max_length)                     
  File "run_seq2seq_no_trainer.py", line 392, in evaluate                                                                      
    model_kwargs = self._prepare_encoder_decoder_kwargs_for_generation(                                                        
  File "/home/sourab/transformers/src/transformers/generation_utils.py", line 525, in _prepare_encoder_decoder_kwargs_for_gener
ation                                                                                                                          
    generated_tokens = unwrapped_model.generate(                                                                               
  File "/home/sourab/dev/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context                
    return func(*args, **kwargs)                                                                                               
  File "/home/sourab/transformers/src/transformers/generation_utils.py", line 1181, in generate                                
    model_kwargs["encoder_outputs"]: ModelOutput = encoder(**encoder_kwargs)                                                   
  File "/home/sourab/dev/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl                     
    model_kwargs = self._prepare_encoder_decoder_kwargs_for_generation(                                                        
  File "/home/sourab/transformers/src/transformers/generation_utils.py", line 525, in _prepare_encoder_decoder_kwargs_for_gener
ation                                                                                                                          
    return forward_call(*input, **kwargs)                                                                                      
  File "/home/sourab/transformers/src/transformers/models/blenderbot/modeling_blenderbot.py", line 736, in forward             
    model_kwargs["encoder_outputs"]: ModelOutput = encoder(**encoder_kwargs)                                                   
  File "/home/sourab/dev/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl                     
    inputs_embeds = self.embed_tokens(input_ids) * self.embed_scale                                                            
  File "/home/sourab/dev/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl                     
    return forward_call(*input, **kwargs)                                                                                      
  File "/home/sourab/transformers/src/transformers/models/blenderbot/modeling_blenderbot.py", line 736, in forward             
    return forward_call(*input, **kwargs)                                                                                      
  File "/home/sourab/dev/lib/python3.8/site-packages/torch/nn/modules/sparse.py", line 158, in forward
    return F.embedding(
  File "/home/sourab/dev/lib/python3.8/site-packages/torch/nn/functional.py", line 2199, in embedding
    inputs_embeds = self.embed_tokens(input_ids) * self.embed_scale
  File "/home/sourab/dev/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/sourab/dev/lib/python3.8/site-packages/torch/nn/modules/sparse.py", line 158, in forward
    return F.embedding(
  File "/home/sourab/dev/lib/python3.8/site-packages/torch/nn/functional.py", line 2199, in embedding
    return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
RuntimeError: The tensor has a non-zero number of elements, but its data is not allocated yet. Caffe2 uses a lazy allocation, s
o you will need to call mutable_data() or raw_mutable_data() to actually allocate memory.
    return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
RuntimeError: The tensor has a non-zero number of elements, but its data is not allocated yet. Caffe2 uses a lazy allocation, s
o you will need to call mutable_data() or raw_mutable_data() to actually allocate memory.
```
5. The error disappears and everything works if I run `model(**dummy_batch)` before prediction loop. The concerned snippet in [run_seq2seq_no_trainer.py](https://github.com/pacman100/fsdp-tests/blob/main/src/modeling/run_seq2seq_no_trainer.py) is shown below:
```python
model.eval()
    if args.val_max_target_length is None:
        args.val_max_target_length = args.max_target_length

    gen_kwargs = {
        "max_length": args.val_max_target_length if args is not None else max_length,
        "num_beams": args.num_beams,
        "min_length": args.val_min_target_length,
        "length_penalty": False,
        "no_repeat_ngram_size": 3,
        "encoder_no_repeat_ngram_size": 3,
        "repetition_penalty": 1.2,
    }
    samples_seen = 0
    for step, batch in enumerate(eval_dataloader):
        # had to run this 1 time at the start of eval loop else was giving device `caffe error`.
        # So, before directly using `model.generate` pass a batch with dummy data through the model
        # uncomment below lines for successful run with FSDP
        # if samples_seen == 0 and accelerator.distributed_type == DistributedType.FSDP:
        #     model(**batch)
        with torch.no_grad():
            unwrapped_model = accelerator.unwrap_model(model)
            generated_tokens = unwrapped_model.generate(
                batch["input_ids"],
                attention_mask=batch["attention_mask"],
                synced_gpus=True,
                **gen_kwargs,
            )
```
On uncommenting the corresponding lines, everything works and below is the output:

![Screenshot 2022-07-29 at 2 50 40 PM](https://user-images.githubusercontent.com/13534540/181728133-297ed562-52fc-4dab-9f8b-3be8ea5f0a37.png)

# Expected Output
No caffe2 allocation error is thrown.

### Versions

Collecting environment information...
PyTorch version: 1.12.0+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-6ubuntu2) 7.5.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8.10 (default, Jun 22 2022, 20:18:18)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.4.0-122-generic-x86_64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration: 
GPU 0: NVIDIA TITAN RTX
GPU 1: NVIDIA TITAN RTX

Nvidia driver version: 510.73.08
cuDNN version: Probably one of the following:
/usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.7.6.5
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.0
[pip3] torch==1.12.0
[pip3] torchaudio==0.12.0
[pip3] torchvision==0.13.0
[conda] Could not collect

cc @ezyang @gchanan @zou3519 @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501

**Comment by pacman100** [comment]
Similar issue faced by another user: https://github.com/huggingface/accelerate/issues/570

**Comment by pacman100** [comment]
@lessw2020, hello,  just pinging you here in case you have the idea or have experienced this issue before.

**Comment by mrshenli** [comment]
cc @zhaojuanmao @rohan-varma

**Comment by lessw2020** [comment]
Hi @pacman100  - sorry I didn't see this earlier. 

To your issue - we've run with T5 and validation loops, with many T5 variations, and did not hit this issue.

However, in reviewing the above it looks like your code is going through some additional HF code before actually passing the inputs to the model for validation, vs while we are using HF T5 model, but then doing everything else in pure PyTorch.

To pin this down a bit more:                                                      

1 - Can you confirm you see this with T5 in your setup above (original text states it happens with all transformers including T5, etc), but most of the specifics seem to revolve around blenderbot.  
Would be a fast test to swap in T5 and see if it shows the same just to make sure if we can reference working T5 code for this issue.

2 - The issue arises immediately upon the first validation mini-batch correct? 
At which point you push in a dummy batch which apparently allocates some embedding related memory for the input, and then things are fine after that. 

I can setup and run blenderbot directly if that would help isolate (vs t5).

Another fast test of this would be to modify your code to work directly with the sharded model in the validation loop, rather than the HF generate function:
ala:
~~~
with torch.no_grad():
        for batch in test_loader:
            for key in batch.keys():
                batch[key] = batch[key].to(local_rank)
            output = model(
                input_ids=batch["source_ids"],
                attention_mask=batch["source_mask"],
                labels=batch["target_ids"],
            )
            ddp_loss[0] += output["loss"].item()  # sum up batch loss
            ddp_loss[1] += len(batch)
           
~~~
You can review the code for T5 here if that helps to compare:
https://github.com/lessw2020/t5_11

If that works then you can work backwards to isolate the issue in the HF utility code.
If it's specific to blenderbot, then please let me know and I can setup blenderbot here in place of T5. 
Thanks for bringing up the issue and look forward to getting it pinned down.

**Comment by piglaker** [comment]
I had the same problem.

**Comment by rohan-varma** [comment]
I think this issue is because FSDP shards the model params, so after you wrap:

FSDP(model)

model will have sharded parameters. In the inference path, it appears that unwrapped_model.generate calls an `encoder` directly, which is not an individually wrapped unit, and we call the forward pass: 

```
 File "/home/sourab/transformers/src/transformers/generation_utils.py", line 1181, in generate                                
    model_kwargs["encoder_outputs"]: ModelOutput = encoder(**encoder_kwargs)                                                   
  File "/home/sourab/dev/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl                     
    model_kwargs = self._prepare_encoder_decoder_kwargs_for_generation(
```

we can tell encoder is not individually wrapped in FSDP because it doesn't appear to go through FSDP codepath. 

I think to help confirm this, can you confirm `model` is FSDP instance and `encoder` is not, and also print out the FSDP wrapping of model? I think the reason `model(batch)` works might be because encoder's params are bubbled up to the top, and the root unit does not reshard after forward, so this sort of works due to an implementation detail.

**Comment by rohan-varma** [comment]
@pacman100 any update from your side? Would it be possible to get a print out of the FSDP wrapped model?

**Comment by SparkJiao** [comment]
> @pacman100 any update from your side? Would it be possible to get a print out of the FSDP wrapped model?

Hi, I have encountered the same problem, and I have tested your comments and what you said is correct. `model` is a FSDP instance while `encoder` is not. So is there any solutions? Thank you very much!

**Comment by rohan-varma** [comment]
We have no plans to support this at the moment, the workaround is to ensure the desired submodule is wrapped in its own FSDP unit. Then as a result, the submodule's forward pass will trigger the appropriate parameter rebuild. 

Added https://github.com/pytorch/pytorch/pull/86343 to clarify this in the docs.

**Comment by lovodkin93** [comment]
> ### 🐛 Describe the bug
> When using FSDP, during inference/evaluation using transformers (gpt2, blenderbot, t5 ...) for generation, i.e., `model.generate()`, caffe2 error is thrown.
> 
> ### Steps to reproduce the error:
> 1. Code is here: [run_seq2seq_no_trainer.py](https://github.com/pacman100/fsdp-tests/blob/main/src/modeling/run_seq2seq_no_trainer.py)
> 2. Using 🤗 Accelerate's FSDP integration with [config.yaml](https://github.com/pacman100/fsdp-tests/blob/main/src/modeling/config.yaml) being below.  Using 2 Nvidia Titan RTX GPUs.
> 
> ```yaml
> compute_environment: LOCAL_MACHINE
> deepspeed_config: {}
> distributed_type: FSDP
> downcast_bf16: 'no'
> fsdp_config:
>   fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
>   fsdp_backward_prefetch_policy: BACKWARD_PRE
>   fsdp_offload_params: false
>   fsdp_sharding_strategy: 1
>   fsdp_state_dict_type: FULL_STATE_DICT
>   fsdp_transformer_layer_cls_to_wrap: BlenderbotDecoderLayer
> machine_rank: 0
> main_process_ip: null
> main_process_port: null
> main_training_function: main
> mixed_precision: 'fp16'
> num_machines: 1
> num_processes: 2
> use_cpu: false
> ```
> 
> 3. Running below [launch](https://github.com/pacman100/fsdp-tests/blob/main/src/modeling/seq2seq_script.sh) command:
> 
> ```shell
> accelerate launch --config_file config.yaml \
>     run_seq2seq_no_trainer.py \
>     --dataset_name "smangrul/MuDoConv" \
>     --max_source_length 128 \
>     --source_prefix "chatbot: " \
>     --max_target_length 64 \
>     --val_max_target_length 64 \
>     --val_min_target_length 20 \
>     --n_val_batch_generations 5 \
>     --n_train 10000 \
>     --n_val 1000 \
>     --pad_to_max_length \
>     --num_beams 10 \
>     --model_name_or_path "facebook/blenderbot-400M-distill" \
>     --per_device_train_batch_size 100 \
>     --per_device_eval_batch_size 50 \
>     --learning_rate 1e-6 \
>     --weight_decay 0.0 \
>     --num_train_epochs 1 \
>     --gradient_accumulation_steps 1 \
>     --num_warmup_steps 100 \
>     --output_dir "/tmp/fsdp_test" \
>     --seed 25 \
>     --logging_steps 100
> ```
> 
> 4. Output with the error and trace stack:
> 
> ```shell
> Traceback (most recent call last):                                                                                             
>   File "run_seq2seq_no_trainer.py", line 911, in <module>                                                                      
>     bleu_score = evaluate(args, model, metric, tokenizer, eval_dataloader, accelerator, config.max_length)                     
>   File "run_seq2seq_no_trainer.py", line 392, in evaluate                                                                      
>     generated_tokens = unwrapped_model.generate(                                                                               
>   File "/home/sourab/dev/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context                
>     main()                                                                                                                     
>   File "run_seq2seq_no_trainer.py", line 840, in main                                                                          
>     return func(*args, **kwargs)                                                                                               
>   File "/home/sourab/transformers/src/transformers/generation_utils.py", line 1181, in generate                                
>     bleu_score = evaluate(args, model, metric, tokenizer, eval_dataloader, accelerator, config.max_length)                     
>   File "run_seq2seq_no_trainer.py", line 392, in evaluate                                                                      
>     model_kwargs = self._prepare_encoder_decoder_kwargs_for_generation(                                                        
>   File "/home/sourab/transformers/src/transformers/generation_utils.py", line 525, in _prepare_encoder_decoder_kwargs_for_gener
> ation                                                                                                                          
>     generated_tokens = unwrapped_model.generate(                                                                               
>   File "/home/sourab/dev/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context                
>     return func(*args, **kwargs)                                                                                               
>   File "/home/sourab/transformers/src/transformers/generation_utils.py", line 1181, in generate                                
>     model_kwargs["encoder_outputs"]: ModelOutput = encoder(**encoder_kwargs)                                                   
>   File "/home/sourab/dev/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl                     
>     model_kwargs = self._prepare_encoder_decoder_kwargs_for_generation(                                                        
>   File "/home/sourab/transformers/src/transformers/generation_utils.py", line 525, in _prepare_encoder_decoder_kwargs_for_gener
> ation                                                                                                                          
>     return forward_call(*input, **kwargs)                                                                                      
>   File "/home/sourab/transformers/src/transformers/models/blenderbot/modeling_blenderbot.py", line 736, in forward             
>     model_kwargs["encoder_outputs"]: ModelOutput = encoder(**encoder_kwargs)                                                   
>   File "/home/sourab/dev/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl                     
>     inputs_embeds = self.embed_tokens(input_ids) * self.embed_scale                                                            
>   File "/home/sourab/dev/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl                     
>     return forward_call(*input, **kwargs)                                                                                      
>   File "/home/sourab/transformers/src/transformers/models/blenderbot/modeling_blenderbot.py", line 736, in forward             
>     return forward_call(*input, **kwargs)                                                                                      
>   File "/home/sourab/dev/lib/python3.8/site-packages/torch/nn/modules/sparse.py", line 158, in forward
>     return F.embedding(
>   File "/home/sourab/dev/lib/python3.8/site-packages/torch/nn/functional.py", line 2199, in embedding
>     inputs_embeds = self.embed_tokens(input_ids) * self.embed_scale
>   File "/home/sourab/dev/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
>     return forward_call(*input, **kwargs)
>   File "/home/sourab/dev/lib/python3.8/site-packages/torch/nn/modules/sparse.py", line 158, in forward
>     return F.embedding(
>   File "/home/sourab/dev/lib/python3.8/site-packages/torch/nn/functional.py", line 2199, in embedding
>     return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
> RuntimeError: The tensor has a non-zero number of elements, but its data is not allocated yet. Caffe2 uses a lazy allocation, s
> o you will need to call mutable_data() or raw_mutable_data() to actually allocate memory.
>     return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
> RuntimeError: The tensor has a non-zero number of elements, but its data is not allocated yet. Caffe2 uses a lazy allocation, s
> o you will need to call mutable_data() or raw_mutable_data() to actually allocate memory.
> ```
> 
> 5. The error disappears and everything works if I run `model(**dummy_batch)` before prediction loop. The concerned snippet in [run_seq2seq_no_trainer.py](https://github.com/pacman100/fsdp-tests/blob/main/src/modeling/run_seq2seq_no_trainer.py) is shown below:
> 
> ```python
> model.eval()
>     if args.val_max_target_length is None:
>         args.val_max_target_length = args.max_target_length
> 
>     gen_kwargs = {
>         "max_length": args.val_max_target_length if args is not None else max_length,
>         "num_beams": args.num_beams,
>         "min_length": args.val_min_target_length,
>         "length_penalty": False,
>         "no_repeat_ngram_size": 3,
>         "encoder_no_repeat_ngram_size": 3,
>         "repetition_penalty": 1.2,
>     }
>     samples_seen = 0
>     for step, batch in enumerate(eval_dataloader):
>         # had to run this 1 time at the start of eval loop else was giving device `caffe error`.
>         # So, before directly using `model.generate` pass a batch with dummy data through the model
>         # uncomment below lines for successful run with FSDP
>         # if samples_seen == 0 and accelerator.distributed_type == DistributedType.FSDP:
>         #     model(**batch)
>         with torch.no_grad():
>             unwrapped_model = accelerator.unwrap_model(model)
>             generated_tokens = unwrapped_model.generate(
>                 batch["input_ids"],
>                 attention_mask=batch["attention_mask"],
>                 synced_gpus=True,
>                 **gen_kwargs,
>             )
> ```
> 
> On uncommenting the corresponding lines, everything works and below is the output:
> 
> ![Screenshot 2022-07-29 at 2 50 40 PM](https://user-images.githubusercontent.com/13534540/181728133-297ed562-52fc-4dab-9f8b-3be8ea5f0a37.png)
> 
> # Expected Output
> No caffe2 allocation error is thrown.
> 
> ### Versions
> Collecting environment information... PyTorch version: 1.12.0+cu102 Is debug build: False CUDA used to build PyTorch: 10.2 ROCM used to build PyTorch: N/A
> 
> OS: Ubuntu 20.04.4 LTS (x86_64) GCC version: (Ubuntu 7.5.0-6ubuntu2) 7.5.0 Clang version: 10.0.0-4ubuntu1 CMake version: version 3.16.3 Libc version: glibc-2.31
> 
> Python version: 3.8.10 (default, Jun 22 2022, 20:18:18) [GCC 9.4.0] (64-bit runtime) Python platform: Linux-5.4.0-122-generic-x86_64-with-glibc2.29 Is CUDA available: True CUDA runtime version: 10.2.89 GPU models and configuration: GPU 0: NVIDIA TITAN RTX GPU 1: NVIDIA TITAN RTX
> 
> Nvidia driver version: 510.73.08 cuDNN version: Probably one of the following: /usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.7.6.5 /usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so.7.6.5 HIP runtime version: N/A MIOpen runtime version: N/A Is XNNPACK available: True
> 
> Versions of relevant libraries: [pip3] numpy==1.23.0 [pip3] torch==1.12.0 [pip3] torchaudio==0.12.0 [pip3] torchvision==0.13.0 [conda] Could not collect
> 
> cc @ezyang @gchanan @zou3519 @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501

@pacman100 
Hey,
I tried what you suggested, but weirdly enough when I run model(**inputs), it instantly reaches GPU capacity of 100%, making the whole pipeline stuck.
Have you encountered such behavior?

**Comment by mostafaelhoushi** [comment]
> I think this issue is because FSDP shards the model params, so after you wrap:
> 
> FSDP(model)
> 
> model will have sharded parameters. In the inference path, it appears that unwrapped_model.generate calls an `encoder` directly, which is not an individually wrapped unit, and we call the forward pass:
> 
> ```
>  File "/home/sourab/transformers/src/transformers/generation_utils.py", line 1181, in generate                                
>     model_kwargs["encoder_outputs"]: ModelOutput = encoder(**encoder_kwargs)                                                   
>   File "/home/sourab/dev/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl                     
>     model_kwargs = self._prepare_encoder_decoder_kwargs_for_generation(
> ```
> 
> we can tell encoder is not individually wrapped in FSDP because it doesn't appear to go through FSDP codepath.
> 
> I think to help confirm this, can you confirm `model` is FSDP instance and `encoder` is not, and also print out the FSDP wrapping of model? I think the reason `model(batch)` works might be because encoder's params are bubbled up to the top, and the root unit does not reshard after forward, so this sort of works due to an implementation detail.

I had a similar error message inferring a non-HF LLM model trained using FSDP, and this explanation solved my problem: it turned out that I was inferring the model with a different number of GPUs from what the model was trained on.

Thanks @rohan-varma !
